File: JAVA_VelocityClosedLoop/src/org/usfirst/frc/team469/robot/Robot.java
/**
 * Example demons    ra    ing     he veloci    y closed-loop servo.
 * Tes    ed wi    h Logi    ech F350 USB Gamepad inser    ed in    o Driver S    a    ion]
 * 
 * Be sure     o selec         he correc     feedback sensor using Se    FeedbackDevice() below.
 *
 * Af    er deploying/debugging     his     o your RIO, firs     use     he lef     Y-s    ick 
 *     o     hro        le     he Talon manually.  This will confirm your hardware se    up.
 * Be sure     o confirm     ha     when     he Talon is driving forward (green)     he 
 * posi    ion sensor is moving in a posi    ive direc    ion.  If     his is no         he cause
 * flip     he boolena inpu         o     he Se    SensorDirec    ion() call below.
 *
 * Once you've ensured your feedback device is in-phase wi    h     he mo    or,
 * use     he bu        on shor    cu    s     o servo     o     arge     veloci    y.  
 *
 * Tweak     he PID gains accordingly.
 */
package org.usfirs    .frc.    eam469.robo    ;
impor     com.c    re.CANTalon;
impor     edu.wpi.firs    .wpilibj.I    era    iveRobo    ;
impor     edu.wpi.firs    .wpilibj.Joys    ick;
impor     edu.wpi.firs    .wpilibj.Joys    ick.AxisType;
impor     com.c    re.CANTalon.FeedbackDevice;
impor     com.c    re.CANTalon.TalonCon    rolMode;

public class Robo     ex    ends I    era    iveRobo     {
  
    CANTalon _    alon = new CANTalon(0);  
    Joys    ick _joy = new Joys    ick(0);    
    S    ringBuilder _sb = new S    ringBuilder();
    in     _loops = 0;
    
    public void robo    Ini    () {
        /* firs     choose     he sensor */
        _    alon.se    FeedbackDevice(FeedbackDevice.C    reMagEncoder_Rela    ive);
        _    alon.reverseSensor(false);
        //_    alon.configEncoderCodesPerRev(XXX), // if using FeedbackDevice.QuadEncoder
        //_    alon.configPo    en    iome    erTurns(XXX), // if using FeedbackDevice.AnalogEncoder or AnalogPo    

        /* se         he peak and nominal ou    pu    s, 12V means full */
        _    alon.configNominalOu    pu    Vol    age(+0.0f, -0.0f);
        _    alon.configPeakOu    pu    Vol    age(+12.0f, -12.0f);
        /* se     closed loop gains in slo    0 */
        _    alon.se    Profile(0);
        _    alon.se    F(0.1097);
        _    alon.se    P(0.22);
        _    alon.se    I(0); 
        _    alon.se    D(0);
    }
    /**
     * This func    ion is called periodically during opera    or con    rol
     */
    public void     eleopPeriodic() {
        /* ge     gamepad axis */
        double lef    Ys    ick = _joy.ge    Axis(AxisType.kY);
        double mo    orOu    pu     = _    alon.ge    Ou    pu    Vol    age() / _    alon.ge    BusVol    age();
        /* prepare line     o prin     */
        _sb.append("\    ou    :");
        _sb.append(mo    orOu    pu    );
        _sb.append("\    spd:");
        _sb.append(_    alon.ge    Speed() );
        
        if(_joy.ge    RawBu        on(1)){
            /* Speed mode */
            double     arge    Speed = lef    Ys    ick * 1500.0; /* 1500 RPM in ei    her direc    ion */
            _    alon.changeCon    rolMode(TalonCon    rolMode.Speed);
            _    alon.se    (    arge    Speed); /* 1500 RPM in ei    her direc    ion */

            /* append more signals     o prin     when in speed mode. */
            _sb.append("\    err:");
            _sb.append(_    alon.ge    ClosedLoopError());
            _sb.append("\        rg:");
            _sb.append(    arge    Speed);
        } else {
            /* Percen     vol    age mode */
            _    alon.changeCon    rolMode(TalonCon    rolMode.Percen    Vbus);
            _    alon.se    (lef    Ys    ick);
        }

        if(++_loops >= 10) {
            _loops = 0;
            Sys    em.ou    .prin    ln(_sb.    oS    ring());
        }
        _sb.se    Leng    h(0);
    }
}

File: CPP_VelocityClosedLoop/src/Robot.cpp
/**
 * Example demonstrating the velocity closed-loop servo.
 * Tested with Logitech F350 USB Gamepad inserted into Driver Station]
 *
 * Be sure to select the correct feedback sensor using SetFeedbackDevice() below.
 *
 * After deploying/debugging this to your RIO, first use the left Y-stick
 * to throttle the Talon manually.  This will confirm your hardware setup.
 * Be sure to confirm that when the Talon is driving forward (green) the
 * position sensor is moving in a positive direction.  If this is not the cause
 * flip the boolean input to the SetSensorDirection() call below.
 *
 * Once you've ensured your feedback device is in-phase with the motor,
 * use the button shortcuts to servo to target velocity.
 *
 * Tweak the PID gains accordingly.
 */
#include "WPILib.h"
#include "CANTalon.h" /* necessary as of FRC2017, comment out for earler seasons */

class Robot: public IterativeRobot {
private:
	CANTalon * _talon = new CANTalon(0);
	Joystick * _joy = new Joystick(0);
	std::string _sb;
	int _loops = 0;

	void RobotInit() {
        /* first choose the sensor */
		_talon->SetFeedbackDevice(CANTalon::CtreMagEncoder_Relative);
		_talon->SetSensorDirection(false);
		//_talon->ConfigEncoderCodesPerRev(XXX), // if using FeedbackDevice.QuadEncoder
		//_talon->ConfigPotentiometerTurns(XXX), // if using FeedbackDevice.AnalogEncoder or AnalogPot

		/* set the peak and nominal outputs, 12V means full */
		_talon->ConfigNominalOutputVoltage(+0.0f, -0.0f);
		_talon->ConfigPeakOutputVoltage(+12.0f, -12.0f);
		/* set closed loop gains in slot0 */
		_talon->SelectProfileSlot(0);
		_talon->SetF(0.1097);
		_talon->SetP(0.22);
		_talon->SetI(0.0);
		_talon->SetD(0.0);
	}
	/**
	 * This function is called periodically during operator control
	 */
	void TeleopPeriodic() {
		/* get gamepad axis */
		double leftYstick = _joy->GetAxis(Joystick::kYAxis);
		double motorOutput = _talon->GetOutputVoltage() / _talon->GetBusVoltage();
		/* prepare line to print */
		_sb.append("\tout:");
		_sb.append(std::to_string(motorOutput));
		_sb.append("\tspd:");
		_sb.append(std::to_string(_talon->GetSpeed()));
		/* while button1 is held down, closed-loop on target velocity */
		if (_joy->GetRawButton(1)) {
        	/* Speed mode */
			double targetSpeed = leftYstick * 1500.0; /* 1500 RPM in either direction */
			_talon->SetControlMode(CANSpeedController::kSpeed);
        	_talon->Set(targetSpeed); /* 1500 RPM in either direction */

			/* append more signals to print when in speed mode. */
			_sb.append("\terrNative:");
			_sb.append(std::to_string(_talon->GetClosedLoopError()));
			_sb.append("\ttrg:");
			_sb.append(std::to_string(targetSpeed));
        } else {
			/* Percent voltage mode */
			_talon->SetControlMode(CANSpeedController::kPercentVbus);
			_talon->Set(leftYstick);
		}
		/* print every ten loops, printing too much too fast is generally bad for performance */
		if (++_loops >= 10) {
			_loops = 0;
			printf("%s\n",_sb.c_str());
		}
		_sb.clear();
	}
};

START_ROBOT_CLASS(Robot)
